Std/Tactic/Case.lean

```diff
@@ -0,0 +1,191 @@
+
+* `case _ : t₁ | _ : t₂ | ... => tac`
+  is equivalent to `(case _ : t₁ => tac); (case _ : t₂ => tac); ...`
+  but with all matching done on the original list of goals, in case pattern elaboration creates
+  new goals.
+  Each goal is consumed as they are matched, so patterns may repeat.
+
+* `case _ : t` will make the matched goal be the first goal.
+  `case _ : t₁ | _ : t₂ | ...` makes the matched goals be the first goals in the given order.
+
+* `case _ : t := _` and `case _ : t := ?m` are the same as `case _ : t` but in the `?m` case the
+  goal tag is changed to `m`.
+  In particular, the goal becomes metavariable `?m`.
+-/
+-- Low priority so that type-free `case` doesn't conflict with core `case`,
+-- though it should be a drop-in replacement.
+syntax (name := casePatt) (priority := low)
+  "case " sepBy1(casePattArg, " | ") (casePattBody)? : tactic
+
+macro_rules
+  | `(tactic| case $[$ps:casePattArg]|* := $t) => `(tactic| case $[$ps:casePattArg]|* => exact $t)
+  | `(tactic| case $[$ps:casePattArg]|*) => `(tactic| case $[$ps:casePattArg]|* => _)
+
+/-- `case' _ : t => tac` is similar to the `case _ : t => tac` tactic,
+but it does not ensure the goal has been solved after applying `tac`,
+nor does it admit the goal if `tac` failed.
+Recall that `case` closes the goal using `sorry` when `tac` fails,
+and the tactic execution is not interrupted. -/
+syntax (name := casePatt') (priority := low)
+  "case' " sepBy1(casePattArg, " | ") casePattTac : tactic
+
+/-- Filter the `mvarIds` by tag. Returns those `MVarId`s that have `tag`
+either as its user name, as a suffix of its user name, or as a prefix of its user name.
+The results are sorted in this order.
+This is like `Lean.Elab.Tactic.findTag?` but it returns all results. -/
+private def filterTag (mvarIds : List MVarId) (tag : Name) : TacticM (List MVarId) := do
+  let gs ← mvarIds.filterMapM fun mvarId => do
+    let userName := (← mvarId.getDecl).userName
+    if tag == userName then
+      return some (0, mvarId)
+    else if tag.isSuffixOf userName then
+      return some (1, mvarId)
+    else if tag.isPrefixOf userName then
+      return some (2, mvarId)
+    else
+      return none
+  -- Insertion sort is a stable sort:
+  let gs := gs.toArray.insertionSort (·.1 < ·.1)
+  return gs |>.map (·.2) |>.toList
+
+/-- Find the first goal among those matching `tag` whose type unifies with `patt`.
+The `renameI` array consists of names to use to rename inaccessibles.
+The `patt` term is elaborated in the context where the inaccessibles have been renamed.
+
+Returns the found goal, goals caused by elaborating `patt`, and the remaining goals. -/
+def findGoalOfPatt (gs : List MVarId)
+    (tag : TSyntax ``binderIdent) (patt? : Option Term) (renameI : TSyntaxArray `Lean.binderIdent) :
+    TacticM (MVarId × List MVarId × List MVarId) :=
+  Term.withoutErrToSorry do
+    let gs ←
+      match tag with
+      | `($tag:ident) => filterTag gs tag.getId
+      | _ => pure gs
```

Comment:
Oops, this `let` shouldn't shadow `gs` since we need to return the remaining goals. I'll review this tomorrow.